libgalois/include/katana/PropertyGraph.h

```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+      Node src, Node dst, const EdgeLabelTy& label) const {
+    // trivial check; can't be connected if degree is 0
+    // TODO(amber): decide whether we check in_degrees
+    // if (degrees_[src] == 0 || in_degrees_[dst] == 0) {
+    // return false;
+    // }
+
+    edge_iterator e_beg = edge_begin(src, label);
+    edge_iterator e_end = edge_end(src, label);
+    // check for degree == 0
+    if (e_beg == e_end) {
+      return false;
+    }
+
+    auto range = internal::make_no_deref_range(e_beg, e_end);
+
+    EdgeComparator comp{this};
+    return std::binary_search(range.begin(), range.end(), dst, comp);
+  }
+
+  /**
+   * Check if vertex src is connected to vertex dst with any edge label
+   *
+   * @param src source node of the edge
+   * @param dst destination node of the edge
+   * @returns true iff the edge exists
+   */
+  bool IsConnected(Node src, Node dst) const {
+    // trivial check; can't be connected if degree is 0
+
+    if (degree(src) == 0ul) {
+      return false;
+    }
+
+    for (const auto& label : GetDistinctEdgeLabels()) {
+      if (IsConnectedWithEdgeLabel(src, dst, label)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+private:
+  using AdjIndexArray = katana::NUMAArray<Edge>;
+  using EdgeDstArray = katana::NUMAArray<Node>;
+  using EdgeLabelArray = katana::NUMAArray<EdgeLabelTy>;
+  using OrigEdgeIDMap = katana::NUMAArray<Edge>;
+
+  // Must invoke SortAllEdgesByDataThenDst() before
+  // calling this function
+  static AdjIndexArray CreateLabeledAdjIndex(
+      const PropertyGraph* pg, const EdgeLabelIndex* edge_label_index,
+      const EdgeShuffleTopology& topo) noexcept;
+
+  EdgeLabelAwareTopology(
+      const PropertyGraph* pg, const EdgeLabelIndex* edge_label_idx,
+      NUMAArray<Edge>&& adj_indices, NUMAArray<Node>&& dests,
+      NUMAArray<Edge>&& orig_edge_ids) noexcept
+      : prop_graph_(pg),
+        edge_label_index_(edge_label_idx),
+        adj_indices_(std::move(adj_indices)),
+        dests_(std::move(dests)),
+        orig_edge_ids_(std::move(orig_edge_ids)) {
+    KATANA_LOG_DEBUG_ASSERT(pg);
+    KATANA_LOG_DEBUG_ASSERT(edge_label_idx);
+
+    KATANA_LOG_DEBUG_ASSERT(
+        adj_indices_.size() == prop_graph_->topology().num_nodes() *
+                                   edge_label_index_->num_unique_labels());
+
+    KATANA_LOG_DEBUG_ASSERT(
+        dests_.size() == prop_graph_->topology().num_edges());
+    KATANA_LOG_DEBUG_ASSERT(dests_.size() == orig_edge_ids_.size());
+  }
+
+  const PropertyGraph* prop_graph_;
+  const EdgeLabelIndex* edge_label_index_;
+  AdjIndexArray adj_indices_;
+  EdgeDstArray dests_;
+  OrigEdgeIDMap orig_edge_ids_;
+};
+
+/**
+ * Provides both out-going and in-coming topology API
+ * Provides Label Aware topology API
+ */
+class KATANA_EXPORT EdgeLabelAwareBiDirTopology
+    : public EdgeLabelAwareTopology {
+  // Inheriting to reuse outgoing topology API
+  //
+  using OutTopoBase = EdgeLabelAwareTopology;
```

Comment:
unless you need to refer to types in a parent class, you can just use `this->foo` to refer to functions in your parent class.